### ISSUE TITLE
"pwhois wa.us" hangs forever

### ISSUE THREAD

**Issue by tobltobs** [joepie91/python-whois#2]
"pwhois wa.us" hangs forever
same result with the api.

**Comment by joepie91** [comment]
I can reproduce the issue here. It appears to be stuck on parsing the registrant data out of the WHOIS response; the specific regex that it's stuck on is the following:

```
"Registrant ID:\s*(?P<handle>.+)\nRegistrant Name:\s*(?P<name>.+)\nRegistrant Organization:\s*(?P<organization>.*)\nRegistrant Address1:\s*(?P<street1>.+)\nRegistrant Address2:\s*(?P<street2>.*)\nRegistrant City:\s*(?P<city>.+)\nRegistrant State\/Province:\s*(?P<state>.+)\nRegistrant Postal Code:\s*(?P<postalcode>.+)\nRegistrant Country:\s*(?P<country>.+)\nRegistrant Country Code:\s*(?P<country_code>.+)\nRegistrant Phone Number:\s*(?P<phone>.+)\nRegistrant Email:\s*(?P<email>.+)\n", # .CO Internet
```

A keyboard interrupt shows this stack trace:

```
Traceback (most recent call last):
  File "./pwhois", line 31, in <module>
    parsed = pythonwhois.parse.parse_raw_whois(data, normalized=True)
  File "/home/sven/projects/python-whois/pythonwhois/parse.py", line 174, in parse_raw_whois
    data["contacts"] = parse_registrants(raw_data)
  File "/home/sven/projects/python-whois/pythonwhois/parse.py", line 506, in parse_registrants
    match = re.search(regex, segment)
  File "/usr/lib64/python2.7/re.py", line 142, in search
    return _compile(pattern, flags).search(string)
KeyboardInterrupt
```

It's a fairly simple regular expression that shouldn't take so long to process, and definitely not for one particular input... I am suspecting this might be a bug in Python's regular expressions library, but I'll look further into it later today.

Are you experiencing this issue for any other domains, or just for `wa.us`? It seems to work fine for me for `whois.us`, which has roughly the same WHOIS data structure (minus the facsimile number).

EDIT: For completeness, the following is the raw WHOIS response:

```
Domain Name:                                 WA.US
Domain ID:                                   D675772-US
Sponsoring Registrar:                        US LOCALITY
Registrar URL (registration services):       www.whois.us
Domain Status:                               serverDeleteProhibited
Domain Status:                               serverHold
Domain Status:                               serverTransferProhibited
Domain Status:                               serverUpdateProhibited
Domain Status:                               inactive
Registrant ID:                               NEUSTAR-US
Registrant Name:                             NEUSTAR
Registrant Organization:                     NEUSTAR
Registrant Address1:                         Loudoun Tech Center
Registrant Address2:                         45980 Center Oak Plaza
Registrant City:                             Sterling
Registrant State/Province:                   VA
Registrant Postal Code:                      20166
Registrant Country:                          United States
Registrant Country Code:                     US
Registrant Phone Number:                     +1.5714345728
Registrant Facsimile Number:                 +1.5714345758
Registrant Email:                            support.us@neustar.us
Registrant Application Purpose:              P5
Registrant Nexus Category:                   C21
Administrative Contact ID:                   NEUSTAR-US
Administrative Contact Name:                 NEUSTAR
Administrative Contact Organization:         NEUSTAR
Administrative Contact Address1:             Loudoun Tech Center
Administrative Contact Address2:             45980 Center Oak Plaza
Administrative Contact City:                 Sterling
Administrative Contact State/Province:       VA
Administrative Contact Postal Code:          20166
Administrative Contact Country:              United States
Administrative Contact Country Code:         US
Administrative Contact Phone Number:         +1.5714345728
Administrative Contact Facsimile Number:     +1.5714345758
Administrative Contact Email:                support.us@neustar.us
Administrative Application Purpose:          P5
Administrative Nexus Category:               C21
Billing Contact ID:                          NEUSTAR-US
Billing Contact Name:                        NEUSTAR
Billing Contact Organization:                NEUSTAR
Billing Contact Address1:                    Loudoun Tech Center
Billing Contact Address2:                    45980 Center Oak Plaza
Billing Contact City:                        Sterling
Billing Contact State/Province:              VA
Billing Contact Postal Code:                 20166
Billing Contact Country:                     United States
Billing Contact Country Code:                US
Billing Contact Phone Number:                +1.5714345728
Billing Contact Facsimile Number:            +1.5714345758
Billing Contact Email:                       support.us@neustar.us
Billing Application Purpose:                 P5
Billing Nexus Category:                      C21
Technical Contact ID:                        NEUSTAR-US
Technical Contact Name:                      NEUSTAR
Technical Contact Organization:              NEUSTAR
Technical Contact Address1:                  Loudoun Tech Center
Technical Contact Address2:                  45980 Center Oak Plaza
Technical Contact City:                      Sterling
Technical Contact State/Province:            VA
Technical Contact Postal Code:               20166
Technical Contact Country:                   United States
Technical Contact Country Code:              US
Technical Contact Phone Number:              +1.5714345728
Technical Contact Facsimile Number:          +1.5714345758
Technical Contact Email:                     support.us@neustar.us
Technical Application Purpose:               P5
Technical Nexus Category:                    C21
Created by Registrar:                        REGISTRY REGISTRAR
Last Updated by Registrar:                   NEULEVELCSR
Last Transferred Date:                       Thu Feb 20 23:46:25 GMT 2003
Domain Registration Date:                    Thu Apr 18 20:42:06 GMT 2002
Domain Expiration Date:                      Fri Apr 17 23:59:59 GMT 2015
Domain Last Updated Date:                    Fri Nov 09 12:00:44 GMT 2012

>>>> Whois database was last updated on: Sat Feb 08 14:49:05 GMT 2014 <<<<

NeuStar, Inc., the Registry Administrator for .US, has collected this
information for the WHOIS database through a .US-Accredited Registrar.
This information is provided to you for informational purposes only and is
designed to assist persons in determining contents of a domain name
registration record in the NeuStar registry database. NeuStar makes this
information available to you "as is" and does not guarantee its accuracy.
By submitting a WHOIS query, you agree that you will use this data only for
lawful purposes and that, under no circumstances will you use this data:
(1) to allow, enable, or otherwise support the transmission of mass
unsolicited, commercial advertising or solicitations via direct mail,
electronic mail, or by telephone; (2) in contravention of any applicable
data and privacy protection laws; or (3) to enable high volume, automated,
electronic processes that apply to the registry (or its systems). Compilation,
repackaging, dissemination, or other use of the WHOIS database in its
entirety, or of a substantial portion thereof, is not allowed without
NeuStar's prior written permission. NeuStar reserves the right to modify or
change these conditions at any time without prior or subsequent notification
of any kind. By executing this query, in any manner whatsoever, you agree to
abide by these terms.

NOTE: FAILURE TO LOCATE A RECORD IN THE WHOIS DATABASE IS NOT INDICATIVE
OF THE AVAILABILITY OF A DOMAIN NAME.

All domain names are subject to certain additional domain name registration
rules.  For details, please visit our site at www.whois.us.
```

**Comment by FichteFoll** [comment]
After 2 hours of investigation (because I'm too lazy to write another write-up):

```
[18:49:14] <FichteFoll> joepie91: I just now realized that the problem is likely because of "\s*.+"
[18:50:26] <FichteFoll> this is because the re engine will test every possible lenght of the \s* match because the following is .+
[18:51:58] <FichteFoll> since you were matching .+ (>+<) anyway you are expecting at least one character anyway
[18:51:34] <joepie91> also, why does this break on wa.us but not on whois.us?
[18:52:15] <FichteFoll> it doesn't break on the other results because the match actually matches there
[18:52:52] <FichteFoll> and it does so with the first try too becausethe re engine would try to match with the most characters with \s*
[18:53:38] <FichteFoll> if the whole regex then doesn't match, it will try to match \s* with one less character
[18:53:52] <FichteFoll> but since you specified a .+ match afterwards, this will work
[18:53:59] <FichteFoll> and it has to test the whole string again
[18:54:38] <FichteFoll> now, there is not only one \s but multiple and it creates a kind of exponential iteration
[18:54:59] <FichteFoll> even though you already know that the expression won't match, the engine itself doesn't
[18:57:59] <FichteFoll> here is the example with the fixed expression: http://hastebin.com/lolevojice.py
[18:59:17] <FichteFoll> you can add "(?:Registrant Facsimile Number:.*)?" after the phone number to make it match
[19:00:46] <FichteFoll> and this is my current write-up that I'm not going to finish now: http://hastebin.com/gurupafaso.txt
```

Full logs: http://hastebin.com/porupehene.xml

**Comment by joepie91** [comment]
In the interest of non-deletionism, here's a (more persistent) log of the latter half of our discussion: http://wire.cryto.net/logs/2014-02-08#T15-05-14

I've implemented a fix for this issue, I'm currently going through testing to ensure it doesn't break anything.

**Comment by joepie91** [comment]
This issue was fixed in 3f7b94687160ac7f5d38d9cb4c88dd695afdb39f.

I've gone ahead and released a new version of python-whois that contains this fix, 2.0.5 (also with added support for some other registrars and WHOIS servers). You can upgrade using `pip install --upgrade pythonwhois` :)
